src/synth.rs

```diff
@@ -0,0 +1,422 @@
+            }
+            for eq in report.new_rws {
+                priors.insert(eq.name.clone(), eq);
+            }
+        }
+
+        Self {
+            params,
+            rng: Pcg64::seed_from_u64(0), // TODO- parameterize?
+            egraph: Default::default(),
+            prior_rws: priors,
+            new_rws: Default::default(),
+            lifting_rws: vec![],
+        }
+    }
+
+    fn enumerate_workload(&self, filename: &str) -> (Vec<RecExpr<L>>, Vec<String>) {
+        let infile = File::open(filename).expect("can't open file");
+        let reader = BufReader::new(infile);
+        let mut terms = vec![];
+        let mut vars: HashSet<String> = HashSet::default();
+        for line in BufRead::lines(reader) {
+            let expr: RecExpr<L> = line.unwrap().parse().unwrap();
+            for node in expr.as_ref() {
+                if let ENodeOrVar::Var(v) = node.clone().to_enode_or_var() {
+                    let mut v = v.to_string();
+                    v.remove(0);
+                    vars.extend(vec![v]);
+                }
+            }
+            terms.push(expr);
+        }
+        (terms, vars.into_iter().collect())
+    }
+
+    fn add_workload(egraph: &mut EGraph<L, SynthAnalysis>, exprs: &[RecExpr<L>]) {
+        for expr in exprs {
+            egraph.add_expr(expr);
+        }
+    }
+
+    fn mk_runner(&self, egraph: EGraph<L, SynthAnalysis>) -> Runner<L, SynthAnalysis, ()> {
+        Runner::default()
+            .with_scheduler(egg::SimpleScheduler)
+            .with_node_limit(usize::MAX)
+            .with_iter_limit(2)
+            .with_time_limit(Duration::from_secs(10))
+            .with_egraph(egraph)
+    }
+
+    fn run_rewrites(
+        &self,
+        mut runner: Runner<L, SynthAnalysis>,
+        rewrites: Vec<&Rewrite<L, SynthAnalysis>>,
+    ) -> (EGraph<L, SynthAnalysis>, HashMap<Id, Vec<Id>>, StopReason) {
+        let ids: Vec<Id> = runner.egraph.classes().map(|c| c.id).collect();
+        runner = runner.run(rewrites);
+        let stop_reason = runner.stop_reason.unwrap();
+
+        let mut found_unions = HashMap::default();
+        for id in ids {
+            let new_id = runner.egraph.find(id);
+            found_unions.entry(new_id).or_insert_with(Vec::new).push(id);
+        }
+
+        runner.egraph.rebuild();
+        (runner.egraph, found_unions, stop_reason)
+    }
+
+    fn apply_unions(&mut self, unions: HashMap<Id, Vec<Id>>) {
+        for ids in unions.values() {
+            if ids.len() > 1 {
+                let first = ids[0];
+                for id in &ids[1..] {
+                    self.egraph.union(first, *id);
+                }
+            }
+        }
+        self.egraph.rebuild();
+    }
+
+    fn cvec_match(&self) -> EqualityMap<L> {
+        // cvecs [𝑎1, . . . , 𝑎𝑛] and [𝑏1, . . . , 𝑏𝑛] match iff:
+        // ∀𝑖. 𝑎𝑖 = 𝑏𝑖 ∨ 𝑎𝑖 = null ∨ 𝑏𝑖 = null and ∃𝑖. 𝑎𝑖 = 𝑏𝑖 ∧ 𝑎𝑖 ≠ null ∧ 𝑏𝑖 ≠ null
```

Comment:
wow this logical formula rendering is so cool.

Reply:
I agree! I didn't expect copy/pasting from the paper to work, but it did!